### ISSUE TITLE
encoder breaks after trying to encode bad input

### ISSUE THREAD

**Issue by jackdoe** [mbarbon/google-protobuf-dynamic#12]
after attempting to encode broken input (null instead of arrayref) the encoder breaks for all subsequent encode() calls
so having schema like:

```
message Person {
  required int32 id = 1;
  repeated SomeSubMessage foo = 6;
  message SomeSubMessage {
    repeated int32 arrayOfInts = 2;
  }
}
```
after attempt to encode `{id => 123, foo => [{arrayOfInts => undef }]}` , all subsequent `encode()` calls produce 0 byte outputs

```
encoded length: 4
Not an array reference when encoding field 'humans.Person.SomeSubMessage.arrayOfInts' at ff20.pl line 31.
encoded length: 0
encoded length: 0
encoded length: 0

```
script to reproduce:

```
use Data::Dumper;
use Google::ProtocolBuffers::Dynamic;
$dynamic = Google::ProtocolBuffers::Dynamic->new;
$dynamic->load_string("person.proto", <<'EOT');
syntax = "proto2";
 
package humans;

message Person {
  required int32 id = 1;
  repeated SomeSubMessage foo = 6;
  message SomeSubMessage {
    repeated int32 arrayOfInts = 2;
  }
}

EOT
 
$dynamic->map({ package => 'humans', prefix => 'Humans' });
 
toggle({id => 123, foo => [{}]});
toggle({id => 123, foo => [{arrayOfInts => undef }]});
toggle({id => 123, foo => [{}]});
toggle({id => 123, foo => [{}]});
toggle({id => 123, foo => [{}]});

sub toggle {
    my ($h) = @_;
    my $encoder = "Humans::Person";
    eval {
        $bytes = $encoder->encode($h);
        print "encoded length: ". length($bytes) . "\n";
        1;
    } or do {
        print $@;
    }
}
```

it works fine if arrayOfInts is in the top object"

```
message Person {
  required int32 id = 1;
  repeated int32 arrayOfInts = 2;
}
```

```
toggle({id => 123 });
toggle({id => 123, arrayOfInts => undef });
toggle({id => 123 });

output:

encoded length: 2
Not an array reference when encoding field 'humans.Person.arrayOfInts' at /home/bnikolov/ff10.pl line 28.
encoded length: 2
```

**Comment by jackdoe** [comment]
same thing happens if i croaks on `Not an hash reference when encoding a humans.Person.SomeSubMessag` if i change from `repeated SomeSubMessage` to `optional SomeSubMessage`

basically:

```
message Person {
  required int32 id = 1;
  repeated int32 arrayOfInts = 2;

  optional SomeSubMessage foo = 6;
  message SomeSubMessage {
    repeated int32 arrayOfInts = 2;
  }
}

toggle({id => 123 });
toggle({id => 123, foo => undef });
toggle({id => 123 });
```

prints:

```
encoded length: 2
Not an hash reference when encoding a humans.Person.SomeSubMessage value at ff10.pl line 34.
encoded length: 0
```

**Comment by mbarbon** [comment]
Fixed in commit 689c3be94747f6cd733cfcfbc1ed3047b06a6f54 and released to CPAN as 0.15.

Thanks!
